docs/GDPs/gdp-0002-regions.rst

```diff
@@ -0,0 +1,290 @@
+Axis Offsets therefore internally hold an offset which is added or subtracted from the indexed point in the axis.
+For example ``I[0] - 2`` is itself an Axis Offset that refers to 2 points before the start of the compute domain in ``I``.
+
+Axis Offsets may be manipulated in Python or in a stencil and can be used as externals in GT4Py to be used in ``region`` subscripts.
+If the external variable is set to ``None``, then any regions using that external are ignored.
+
+Region Specification
+++++++++++++++++++++
+
+``region`` is a keyword in GT4Py that, when subscripted using slices of axis offsets, defines the restricted computation.
+These form the arguments to ``with horizontal()``.
+
+As an example, ``region[I[0], :]`` specifies a restriction to the first compute point on the ``I`` axis, and no restriction in the ``J`` axis.
+``I[0]`` is a single point, so when converted to a slice is still the single point.
+The ``J`` axis simply has ``:``, which is an unrestricted slice, which GT4Py interprets as an unrestricted axis (behaves normally).
+
+The previous example introduced a key element of regional computation: There must not only be a way of specifying axis offsets outside the compute domain, but slices that extend to infinity in each direction (or alternatively, unrestricted endpoints of axes).
+GTScript interprets an unrestricted start or stop element as extending to infinity (or, unrestricted).
+This is useful in the case when writing a stencil and requiring that an edge condition be made without knowing how far the statements needs to be extended.
+For example:
+
+.. code-block:: python
+
+    with computation(PARALLEL), interval(...):
+        with horizontal(region[:I[0], :]):
+            u = 0
+
+This will set ``u=0`` in all extended computation points to the left of the compute domain.
+
+Examples of this are shown in the image below.
+The blue line shows the compute domain along the ``I`` axis, and two examples of region axis slices are shown in red.
+
+.. image:: _static/gdp-0002/axis_offsets.svg
+
+Execution
++++++++++
+
+Another key feature to remember when using regions is that these should be thought of as specifying specialized computation at points.
+These are therefore not guaranteed to execute, except where inside the compute domain.
+The statements inside a block with ``region[:I[0]-1, :]`` will only execute where the outputs from that block are necessary to compute something else with an extent.
+For example, the following will execute
+
+.. code:: python
+
+    with computation(PARALLEL), interval(...):
+        with horizontal(region[I[0]-1, :]):
+            field_in = 0.0
+        field_out = field_in[-1, 0, 0] + field_in[0, 0, 0]
+
+since the ``field_in`` value at ``I[0]-1`` is being consumed to compute a value of an output field inside the compute domain.
+If this used ``I[0]-2``, the code would be ignored.
```

Comment:
Is this a safe assumption? What if the user wants that computation to be visible outside the stencil specification? I guess in that case the computation should have been expressed separately. Another way of saying the same is asking if this computation would be executed:

```
with computation(PARALLEL), interval(...):
    with horizontal(region[I[0]-1, :]):
        field = 0.0
```

I can see why you want to optimize the code out, and it's fine with me if this is the semantic we always want

Reply:
Yes, I would say this is correct. If the user wants it exposed outside the computation then they would have to write the region differently: `region[I[0], :]`, or define an external `ROW=I[0]`, and use it in `region[ROW, :]`.

The idea behind this is that GT4Py stencils are not a complete picture of an application. Talking only about stencils can hide the reality that these are defined on an application grid space, if the operator needs to be specialized on different parts of the grid. Say, for example you want to apply a different finite difference stencil on edges or at corners, which is required for cubed-sphere grid. In this context, the code that goes in regions allows you to define computation with respect to the application grid.

I see two different semantics:
1. Regions define computation no matter what, breaking the natural extent analysis from a single stencil compute domain. I do not yet see through how extent analysis for this method would work. Perhaps every region would spawn a new extent analysis?
2. Define specialized computation at points. If fields indices set in this block are used later, expose this calculation.

The frontend feature currently exposes option (2).

Reply:
Ok, I see. As long as the semantic is clear I have no objections, I wanted to make sure we were aware of things. Thanks!